### DISCUSSION TITLE
Bad Usage of Commands and Events

### DISCUSSION THREAD

**Discussion by cloud303-cholden** [bevyengine/bevy#11732] [Q&A] [answered]
I'm wanting to trigger `EntityCommands` from an `Event` with an `Entity` field. Is this an anti-pattern in Bevy? The reason I ask is because I've started to run into a lot of `NoSuchEntity` errors, and I'm realizing that it's probably because commands are cleared every frame while events can persist an extra frame. If so, are there workarounds or alternatives? I want to ensure the events are processed exactly once, so flushing unprocessed events at the end of the frame isn't a good option for me. I think all of my affected code could be updated to use pure commands, but using events has made it really easy to isolate the logic for triggering/handling/chaining effects to entities.

**Comment by Aeropolian** [comment, marked as answer]
While events can linger over a frame this is only to ensure every system will process those events. Every event will only be seen once by each system, it just may be a frame late if the system already ran before the event was sent. I have encountered similar issues with those events crossing frame boundaries under certain circumstances. Typically, in places where the entity may be destroyed before the the event is processed. You may need to implement checking to see if the entity exists before attempting to handle the event.

Since you are doing events associated with an `Entity`, it may be helpful to use the 3rd party library [`bevy_eventlistener`](https://github.com/aevyrie/bevy_eventlistener). Based on my understanding, the library is built specifically for entity events and has additional functionality which you may find of use in how and when those are propagated. However, you will need to refer to its documentation to see if it fits your use case.

**Reply by cloud303-cholden** [reply]
Did some testing, and this was exactly what I needed. Thanks!
